cli.go

```diff
@@ -0,0 +1,212 @@
+package jwtauth
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/signal"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/hashicorp/vault/api"
+)
+
+const defaultMount = "oidc"
+const defaultPort = "8300"
+
+type CLIHandler struct{}
+
+type loginResp struct {
+	secret *api.Secret
+	err    error
+}
+
+func (h *CLIHandler) Auth(c *api.Client, m map[string]string) (*api.Secret, error) {
+	// handle ctrl-c while waiting for the callback
+	sigintCh := make(chan os.Signal, 1)
+	signal.Notify(sigintCh, os.Interrupt)
+	defer signal.Stop(sigintCh)
+
+	doneCh := make(chan loginResp)
+
+	mount, ok := m["mount"]
+	if !ok {
+		mount = defaultMount
+	}
+
+	port, ok := m["port"]
+	if !ok {
+		port = defaultPort
+	}
+
+	role := m["role"]
+	if role == "" {
+		return nil, errors.New("a 'role' must be specified")
+	}
+
+	authURL, err := fetchAuthURL(c, role, mount, port)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up callback handler
+	http.HandleFunc(fmt.Sprintf("/ui/vault/auth/%s/oidc/callback", mount), func(w http.ResponseWriter, req *http.Request) {
+		var response string
+
+		query := req.URL.Query()
+		code := query.Get("code")
+		state := query.Get("state")
+		data := map[string][]string{
+			"code":  {code},
+			"state": {state},
+		}
+
+		secret, err := c.Logical().ReadWithData(fmt.Sprintf("auth/%s/oidc/callback", mount), data)
+		if err != nil {
+			summary, detail := parseError(err)
+			response = errorHTML(summary, detail)
+		} else {
+			response = successHTML
+		}
+
+		w.Write([]byte(response))
+		doneCh <- loginResp{secret, err}
+	})
+
+	listener, err := net.Listen("tcp", ":"+port)
+	if err != nil {
+		return nil, err
+	}
+
+	// Open the default browser to the callback URL.
+	fmt.Fprintf(os.Stderr, "Complete the login via your OIDC provider. Launching browser to:\n\n    %s\n\n\n", authURL)
+	if err := openURL(authURL); err != nil {
+		fmt.Fprintf(os.Stderr, "Error attempting to automatically open browser: '%s'.\nPlease visit the authorization URL manually.", err)
+	}
+
+	// Start local server
+	go func() {
+		err := http.Serve(listener, nil)
+		if err != nil && err != http.ErrServerClosed {
+			doneCh <- loginResp{nil, err}
+		}
+	}()
+
+	// Wait for either the callback to finish or SIGINT to be received
+	select {
+	case s := <-doneCh:
+		return s.secret, s.err
+	case <-sigintCh:
+		return nil, errors.New("interrupted")
+	}
+}
+
+func fetchAuthURL(c *api.Client, role, mount, port string) (string, error) {
+	data := map[string]interface{}{
+		"role":         role,
+		"redirect_uri": fmt.Sprintf("http://localhost:%s/ui/vault/auth/%s/oidc/callback", port, mount),
```

Comment:
Is there a reason for this path to be so long? My worry is that it won't actually match the path used by a user (it doesn't, for instance, take into account namespaces) so will be confusing. It could just be `localhost:%s/oidc/callback` right?

Speaking of which -- I assume this will support namespaces via simply using the API client but may be good to check.

Reply:
For the purposes of the CLI listener, it could anything. It's currently set up to match the pattern used by the UI so an operator can set up two redirect_uris that are the same except for the host portion. e.g. to provide both UI and CLI login ability, both of these would need to be in both the provider config and the Vault config.

https://vault-server.bigcorp.com:8200/ui/vault/auth/jwt/oidc/callback
http://localhost:8300/ui/vault/auth/jwt/oidc/callback

Does the similarity help, hurt? I can go either way on this one. Looking at them together, something really simple/short for localhost is probably better.

Yes I was relying on the API client to make NS work but it verify it is.

Reply:
Replaced with `localhost:%s/oidc/callback`